src/mongocrypt-crypto.c

```diff
@@ -1658,3 +1658,35 @@ _mongocrypt_fle2_do_decryption (_mongocrypt_crypto_t *crypto,
 
    return true;
 }
+
+bool
+_mongocrypt_random_int64 (_mongocrypt_crypto_t *crypto,
+                          int64_t exclusive_upper_bound,
+                          int64_t *out,
+                          mongocrypt_status_t *status)
+{
+   if (exclusive_upper_bound <= 0) {
+      CLIENT_ERR ("Expected exclusive_upper_bound > 0");
+      return false;
+   }
+
+   _mongocrypt_buffer_t rand_i64_buf;
+   _mongocrypt_buffer_init (&rand_i64_buf);
+   _mongocrypt_buffer_resize (&rand_i64_buf, (uint32_t) sizeof (int64_t));
+
+   if (!_mongocrypt_random (crypto, &rand_i64_buf, rand_i64_buf.len, status)) {
+      _mongocrypt_buffer_cleanup (&rand_i64_buf);
+      return false;
+   }
+
+   int64_t rand_i64;
+   memcpy (&rand_i64, rand_i64_buf.data, rand_i64_buf.len);
+
+   /* Zero the leading bit to ensure rand_i64 is non-negative. */
+   rand_i64 &= (~(1ull << 63));
+
+   *out = rand_i64 % exclusive_upper_bound;
```

Comment:
Do we want strongly random numbers? This will introduce a subtle bias for non-powers-of-two `exclusive_upper_bound`.

Reply:
Scratch the divide-by-zero, since that's checked, but still might overflow

Reply:
Great question. I was not aware about the modulo bias.

I am not sure if strongly random numbers are a requirement. The server uses a [secure random number generator](https://github.com/mongodb/mongo/blob/0ba7cf78b7d59497eafd38b5aafeac29e8484cf1/src/mongo/platform/random.h#L185). On Unix it is sourced from `/dev/urandom` into `std::uniform_int_distribution`.

@markbenvenuto is strongly random numbers a requirement for contention counter generation?

One straightforward looking option may be to implement something like [arc4_random_uniform](https://github.com/openbsd/src/blob/2207c4325726fdc5c4bcd0011af0fdf7d3dab137/lib/libc/crypt/arc4random_uniform.c#L33).

Reply:
Updated to add a `_mongocrypt_random_uint64` based on the linked implementation of `arc4random_uniform`.